app/code/OleksiiBezpoiasnyi/RegularCustomer/Observer/GuestToCustomerDiscountRequests.php

```diff
@@ -0,0 +1,64 @@
+<?php
+declare(strict_types=1);
+
+namespace OleksiiBezpoiasnyi\RegularCustomer\Observer;
+
+use Magento\Customer\Model\Customer;
+use Magento\Framework\Event\Observer;
+use OleksiiBezpoiasnyi\RegularCustomer\Model\DiscountRequest;
+use OleksiiBezpoiasnyi\RegularCustomer\Model\ResourceModel\DiscountRequest\Collection;
+
+class GuestToCustomerDiscountRequests implements \Magento\Framework\Event\ObserverInterface
+{
+    private \Magento\Customer\Model\Session $customerSession;
+
+    private \OleksiiBezpoiasnyi\RegularCustomer\Model\ResourceModel\DiscountRequest\CollectionFactory $collectionFactory;
+
+    private \Magento\Framework\DB\TransactionFactory $transactionFactory;
+
+    /**
+     * GuestToCustomerDiscountRequests constructor.
+     * @param \Magento\Customer\Model\Session $customerSession
+     * @param \OleksiiBezpoiasnyi\RegularCustomer\Model\ResourceModel\DiscountRequest\CollectionFactory $collectionFactory
+     * @param \Magento\Framework\DB\TransactionFactory $transactionFactory
+     */
+    public function __construct(
+        \Magento\Customer\Model\Session $customerSession,
+        \OleksiiBezpoiasnyi\RegularCustomer\Model\ResourceModel\DiscountRequest\CollectionFactory $collectionFactory,
+        \Magento\Framework\DB\TransactionFactory $transactionFactory
+    ) {
+        $this->customerSession = $customerSession;
+        $this->collectionFactory = $collectionFactory;
+        $this->transactionFactory = $transactionFactory;
+    }
+
+    /**
+     * @param Observer $observer
+     * @return void
+     * @throws \Exception
+     */
+    public function execute(Observer $observer): void
+    {
+
+        /** @var Customer $customer */
+        $customer = $observer->getData('customer');
+        $customerId = $customer->getId();
+        $guestEmail = $this->customerSession->getData('guest_email');
+        $guestProductList = $this->customerSession->getData('product_list');
+
+        /** @var Collection $collection */
+        $collection = $this->collectionFactory->create();
+        $collection->addFieldToFilter('product_id', ['in' => $guestProductList])
+            ->addFieldToFilter('customer_id', ['null' => true])
+            ->addFieldToFilter('email', $guestEmail);
```

Comment:
@AlexeyBezpoyasny There is no backend validation of the email in class `\OleksiiBezpoiasnyi\RegularCustomer\Controller\LoyaltyProgram\Registration`. It is possible to disable JS and send a request without email. After being logged all requests for the product without email will be assigned to that customer.
1) There must be an email validation in the resource model. It can be added to the `_beforeSave()` method. Saving a request without email should not be possible.
2) Must anyway check that customer emails is at least not empty. Can add email validation to the observer as well.